MFA Remember Device Fails with MariaDb due to keyword conflict

CAS 5.1.4 with Tomcat 7, Java 8 and MariaDb 10

Columns for Remember this device conflict with MariaDb Keywords, specifically `Date` and `Key` the column names are not backticked so SQL fails with SQL Syntax error.

Caused by: java.sql.SQLException: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'key, name, principal) values ('2017-10-11', '136.244.201.1@Mozilla/5.0 (Macintos' at line 1
Query is: insert into MultifactorAuthenticationTrustRecord (date, geography, key, name, principal) values (?, ?, ?, ?, ?), parameters ['2017-10-11','136.244.201.1@Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36','atilling@136.244.201.1@Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36','My Dekstop','atilling']
	at org.mariadb.jdbc.internal.util.LogQueryTool.exceptionWithQuery(LogQueryTool.java:146) ~[mariadb-java-client-2.1.2.jar:?]
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:217) ~[mariadb-java-client-2.1.2.jar:?]
	at sun.reflect.GeneratedMethodAccessor195.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_141]

Issue is discussed at length here: 
https://groups.google.com/a/apereo.org/forum/#!msg/cas-user/3Je3HiqdAhU/2oJZbbGvBwAJ

Overlay is here:
https://github.com/ConnCollege/cas-5/tree/MFAConfig

--- a/support/cas-server-support-trusted-mfa-jdbc/src/main/java/org/apereo/cas/trusted/authentication/storage/JpaMultifactorAuthenticationTrustStorage.java
+++ b/support/cas-server-support-trusted-mfa-jdbc/src/main/java/org/apereo/cas/trusted/authentication/storage/JpaMultifactorAuthenticationTrustStorage.java
@@ -32,11 +32,11 @@ public class JpaMultifactorAuthenticationTrustStorage extends BaseMultifactorAut
     private EntityManager entityManager;
 
     @Override
-    public void expire(final String key) {
+    public void expire(final String recordKey) {
         try {
-            final int count = this.entityManager.createQuery("DELETE FROM " + TABLE_NAME + " r where r.key = :key",
+            final int count = this.entityManager.createQuery("DELETE FROM " + TABLE_NAME + " r where r.recordKey = :recordKey",
                     MultifactorAuthenticationTrustRecord.class)
-                    .setParameter("key", key)
+                    .setParameter("recordKey", recordKey)
                     .executeUpdate();
             LOGGER.info("Found and removed [{}] records", count);
         } catch (final NoResultException e) {
@@ -47,9 +47,9 @@ public class JpaMultifactorAuthenticationTrustStorage extends BaseMultifactorAut
     @Override
     public void expire(final LocalDate onOrBefore) {
         try {
-            final int count = this.entityManager.createQuery("DELETE FROM " + TABLE_NAME + " r where r.date < :date",
+            final int count = this.entityManager.createQuery("DELETE FROM " + TABLE_NAME + " r where r.recordDate < :recordDate",
                     MultifactorAuthenticationTrustRecord.class)
-                    .setParameter("date", onOrBefore)
+                    .setParameter("recordDate", onOrBefore)
                     .executeUpdate();
             LOGGER.info("Found and removed [{}] records", count);
         } catch (final NoResultException e) {
@@ -61,8 +61,8 @@ public class JpaMultifactorAuthenticationTrustStorage extends BaseMultifactorAut
     public Set<MultifactorAuthenticationTrustRecord> get(final LocalDate onOrAfterDate) {
         try {
             final List<MultifactorAuthenticationTrustRecord> results =
-                    this.entityManager.createQuery("SELECT r FROM " + TABLE_NAME + " r where r.date >= :date",
-                            MultifactorAuthenticationTrustRecord.class).setParameter("date", onOrAfterDate).getResultList();
+                    this.entityManager.createQuery("SELECT r FROM " + TABLE_NAME + " r where r.recordDate >= :recordDate",
+                            MultifactorAuthenticationTrustRecord.class).setParameter("recordDate", onOrAfterDate).getResultList();
             return new HashSet<>(results);
         } catch (final NoResultException e) {
             LOGGER.info("No trusted authentication records could be found for [{}]", onOrAfterDate);

--- a/support/cas-server-support-trusted-mfa/src/main/java/org/apereo/cas/trusted/authentication/api/MultifactorAuthenticationTrustRecord.java
+++ b/support/cas-server-support-trusted-mfa/src/main/java/org/apereo/cas/trusted/authentication/api/MultifactorAuthenticationTrustRecord.java
@@ -40,10 +40,10 @@ public class MultifactorAuthenticationTrustRecord implements Comparable<Multifac
 
     @Column(nullable = false, columnDefinition = "TIMESTAMP")
     @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
-    private LocalDate date;
+    private LocalDate recordDate;
 
-    @Column(length = 255, updatable = true, insertable = true, nullable = false)
-    private String key;
+    @Column(length = 500, updatable = true, insertable = true, nullable = false)
+    private String recordKey;
 
     @Column(length = 255, updatable = true, insertable = true, nullable = false)
     private String name;
@@ -52,12 +52,12 @@ public class MultifactorAuthenticationTrustRecord implements Comparable<Multifac
         this.id = System.currentTimeMillis();
     }
 
-    public String getKey() {
-        return key;
+    public String getRecordKey() {
+    	return recordKey;
     }
 
-    public void setKey(final String id) {
-        this.key = id;
+    public void setRecordKey(final String id) {
+        this.recordKey = id;
     }
 
     public String getPrincipal() {
@@ -76,12 +76,12 @@ public class MultifactorAuthenticationTrustRecord implements Comparable<Multifac
         this.geography = geography;
     }
 
-    public LocalDate getDate() {
-        return date;
+    public LocalDate getRecordDate() {
+        return recordDate;
     }
 
-    public void setDate(final LocalDate date) {
-        this.date = date;
+    public void setRecordDate(final LocalDate date) {
+        this.recordDate = date;
     }
 
     @Override
@@ -99,8 +99,8 @@ public class MultifactorAuthenticationTrustRecord implements Comparable<Multifac
         return new EqualsBuilder()
                 .append(this.principal, rhs.principal)
                 .append(this.geography, rhs.geography)
-                .append(this.date, rhs.date)
-                .append(this.key, rhs.key)
+                .append(this.recordDate, rhs.recordDate)
+                .append(this.recordKey, rhs.recordKey)
                 .append(this.name, rhs.name)
                 .isEquals();
     }
@@ -118,8 +118,8 @@ public class MultifactorAuthenticationTrustRecord implements Comparable<Multifac
         return new HashCodeBuilder()
                 .append(principal)
                 .append(geography)
-                .append(date)
-                .append(key)
+                .append(recordDate)
+                .append(recordKey)
                 .append(name)
                 .toHashCode();
     }
@@ -129,7 +129,7 @@ public class MultifactorAuthenticationTrustRecord implements Comparable<Multifac
         return new ToStringBuilder(this, ToStringStyle.NO_CLASS_NAME_STYLE)
                 .append("principal", principal)
                 .append("geography", geography)
-                .append("date", date)
+                .append("recordDate", recordDate)
                 .append("name", name)
                 .toString();
     }
@@ -143,7 +143,7 @@ public class MultifactorAuthenticationTrustRecord implements Comparable<Multifac
      */
     public static MultifactorAuthenticationTrustRecord newInstance(final String principal, final String geography) {
         final MultifactorAuthenticationTrustRecord r = new MultifactorAuthenticationTrustRecord();
-        r.setDate(LocalDate.now());
+        r.setRecordDate(LocalDate.now());
         r.setPrincipal(principal);
         r.setGeography(geography);
         r.setName(principal.concat("-").concat(LocalDate.now().toString()).concat("-").concat(geography));
@@ -152,6 +152,6 @@ public class MultifactorAuthenticationTrustRecord implements Comparable<Multifac
 
     @Override
     public int compareTo(final MultifactorAuthenticationTrustRecord o) {
-        return this.date.compareTo(o.getDate());
+        return this.recordDate.compareTo(o.getRecordDate());
     }
 }

--- a/support/cas-server-support-trusted-mfa/src/main/java/org/apereo/cas/trusted/authentication/storage/BaseMultifactorAuthenticationTrustStorage.java
+++ b/support/cas-server-support-trusted-mfa/src/main/java/org/apereo/cas/trusted/authentication/storage/BaseMultifactorAuthenticationTrustStorage.java
@@ -1,4 +1,4 @@
-package org.apereo.cas.trusted.authentication.storage;
+	package org.apereo.cas.trusted.authentication.storage;
 
 import org.apache.commons.lang3.StringUtils;
 import org.apereo.cas.CipherExecutor;
@@ -33,7 +33,7 @@ public abstract class BaseMultifactorAuthenticationTrustStorage implements Multi
     @Override
     public MultifactorAuthenticationTrustRecord set(final MultifactorAuthenticationTrustRecord record) {
         LOGGER.debug("Stored authentication trust record for [{}]", record);
-        record.setKey(generateKey(record));
+        record.setRecordKey(generateKey(record));
         return setInternal(record);
     }
 
@@ -41,10 +41,10 @@ public abstract class BaseMultifactorAuthenticationTrustStorage implements Multi
     public Set<MultifactorAuthenticationTrustRecord> get(final String principal, final LocalDate onOrAfterDate) {
         final Set<MultifactorAuthenticationTrustRecord> res = get(principal);
         res.removeIf(entry -> {
-            if (entry.getDate().isBefore(onOrAfterDate)) {
+            if (entry.getRecordDate().isBefore(onOrAfterDate)) {
                 return true;
             }
-            final String decodedKey = this.cipherExecutor.decode(entry.getKey());
+            final String decodedKey = this.cipherExecutor.decode(entry.getRecordKey());
             final String currentKey = MultifactorAuthenticationTrustUtils.generateKey(entry);
             if (StringUtils.isBlank(decodedKey)) {
                 return true;

--- a/support/cas-server-support-trusted-mfa/src/main/java/org/apereo/cas/trusted/authentication/storage/InMemoryMultifactorAuthenticationTrustStorage.java
+++ b/support/cas-server-support-trusted-mfa/src/main/java/org/apereo/cas/trusted/authentication/storage/InMemoryMultifactorAuthenticationTrustStorage.java
@@ -34,14 +34,14 @@ public class InMemoryMultifactorAuthenticationTrustStorage extends BaseMultifact
         final Set<MultifactorAuthenticationTrustRecord> results = storage.asMap()
                 .values()
                 .stream()
-                .filter(entry -> entry.getDate().isEqual(onOrBefore) || entry.getDate().isBefore(onOrBefore))
+                .filter(entry -> entry.getRecordDate().isEqual(onOrBefore) || entry.getRecordDate().isBefore(onOrBefore))
                 .sorted()
                 .distinct()
                 .collect(Collectors.toSet());
 
         LOGGER.info("Found [{}] expired records", results.size());
         if (!results.isEmpty()) {
-            results.forEach(entry -> storage.invalidate(entry.getKey()));
+            results.forEach(entry -> storage.invalidate(entry.getRecordKey()));
             LOGGER.info("Invalidated and removed [{}] expired records", results.size());
         }
     }
@@ -52,7 +52,7 @@ public class InMemoryMultifactorAuthenticationTrustStorage extends BaseMultifact
         return storage.asMap()
                 .values()
                 .stream()
-                .filter(entry -> entry.getDate().isEqual(onOrAfterDate) || entry.getDate().isAfter(onOrAfterDate))
+                .filter(entry -> entry.getRecordDate().isEqual(onOrAfterDate) || entry.getRecordDate().isAfter(onOrAfterDate))
                 .sorted()
                 .distinct()
                 .collect(Collectors.toSet());
@@ -72,7 +72,7 @@ public class InMemoryMultifactorAuthenticationTrustStorage extends BaseMultifact
 
     @Override
     public MultifactorAuthenticationTrustRecord setInternal(final MultifactorAuthenticationTrustRecord record) {
-        this.storage.put(record.getKey(), record);
+        this.storage.put(record.getRecordKey(), record);
         return record;
     }
 }

